lib/ResultExtractor.php

```diff
@@ -62,64 +63,27 @@ public function extract($data, array $facetBuilders = [], array $aggregations =
-            // visit stage in NativeQueryConverter.
-            $facetBuildersById = [];
-            foreach ($facetBuilders as $facetBuilder) {
-                $facetBuildersById[spl_object_hash($facetBuilder)] = $facetBuilder;
-            }
-
-            foreach ($data->facet_counts as $facetCounts) {
-                foreach ($facetCounts as $field => $facet) {
-                    if (empty($facetBuildersById[$field])) {
-                        @trigger_error(
-                            'Not setting id of field using FacetFieldVisitor::visitBuilder will not be supported in 2.0'
-                            . ', as it makes it impossible to exactly identify which facets belongs to which builder.'
-                            . "\nMake sure to adapt your visitor for the following field: ${field}"
-                            . "\nExample: 'facet.field' => \"{!ex=dt key=\${id}}${field}\",",
-                            E_USER_DEPRECATED);
-                    }
-
-                    $result->facets[] = $this->facetBuilderVisitor->mapField(
-                        $field,
-                        (array)$facet,
-                        isset($facetBuildersById[$field]) ? $facetBuildersById[$field] : null
-                    );
-                }
-            }
-        }
-
-        $aggregationsResults = [];
-        foreach ($aggregations as $aggregation) {
-            $name = $aggregation->getName();
-
-            if (isset($data->facets->{$name})) {
-                $aggregationsResults[] = $this->aggregationResultExtractor->extract(
-                    $aggregation,
-                    $languageFilter,
-                    $data->facets->{$name}
-                );
-            }
-        }
-
-        $result->aggregations = new AggregationResultCollection($aggregationsResults);
+        $result->facets = $this->extractFacets($data, $facetBuilders, $languageFilter);
+        $result->aggregations = $this->extractAggregations($data, $aggregations, $languageFilter);
 
         foreach ($data->response->docs as $doc) {
-            $searchHit = new SearchHit(
-                [
-                    'score' => $doc->score,
-                    'index' => $this->getIndexIdentifier($doc),
-                    'matchedTranslation' => $this->getMatchedLanguageCode($doc),
-                    'valueObject' => $this->extractHit($doc),
-                ]
-            );
-            $result->searchHits[] = $searchHit;
+            $result->searchHits[] = $this->extractSearchHit($doc, $languageFilter);
         }
 
         return $result;
     }
 
+    /**
+     * Extracts value object from $hit returned by Solr backend.
+     *
+     * Needs to be implemented by the concrete ResultExtractor.
+     *
+     * @param mixed $hit
+     *
+     * @return \eZ\Publish\API\Repository\Values\ValueObject
+     */
+    abstract public function extractHit($hit);
```

Comment:
Can this be strictly typed?

```suggestion
    abstract public function extractHit($hit): ValueObject;
```

Reply:
This would be BC break. Technically this class as is not part of the API/SPI so shouldn't be extended but reality is that in many cases there is no other choice.

Reply:
> This would be BC break. Technically this class as is not part of the API/SPI so shouldn't be extended but reality is that in many cases there is no other choice.

This is not supported.

Reply:
How would this be a BC break if this is a method (abstract) that you just extracted?

Reply:
Nvm, `git` trolled me and split the diff in a strange place.

:+1: